tests/lib/c_lib/src/test_sqrt.c

```diff
@@ -0,0 +1,168 @@
+#include <zephyr/ztest.h>
+
+
+#ifndef NAN
+#define NAN	(__builtin_nansf(""))
+#endif
+
+#ifndef NANF
+#define NANF	(__builtin_nans(""))
+#endif
+
+#ifndef INF
+#define INF	(__builtin_inf())
+#endif
+
+#ifndef INFF
+#define INFF	(__builtin_inff())
+#endif
+
+static float test_floats[] = {
+	1.0f, 2.0f, 3.0f, 4.0f,
+	5.0f, 6.0f, 7.0f, 8.0f, 9.0f,	/* numbers across the decade */
+	3.14159265359f, 2.718281828f,	/* irrational numbers pi and e */
+	123.4f,	0.025f,	0.10f,	1.875f	/* numbers with infinite */
+					/* repeating binary representation */
+	};
+#define	NUM_TEST_FLOATS	(sizeof(test_floats)/sizeof(float))
+
+	static double test_doubles[] = {
+		1.0, 2.0, 3.0, 4.0,
+		5.0, 6.0, 7.0, 8.0, 9.0,	/* numbers across the decade */
+		3.14159265359, 2.718281828,	/* irrational numbers pi and e */
+		123.4,	0.025,	0.10,	1.875	/* numbers with infinite */
+						/* repeating binary representationa */
+};
+#define	NUM_TEST_DOUBLES	(sizeof(test_floats)/sizeof(float))
+
+#ifndef	isinf
+static int isinf(double x)
+{
+	union { uint64_t u; double d; } ieee754;
+	ieee754.d = x;
+	ieee754.u &= ~0x8000000000000000;	/* ignore the sign */
+	return ((ieee754.u >> 52) == 0x7FF) &&
+		((ieee754.u & 0x000fffffffffffff) == 0);
+}
+#endif
+
+#ifndef	isnan
+static int isnan(double x)
+{
+	union { uint64_t u; double d; } ieee754;
+	ieee754.d = x;
+	ieee754.u &= ~0x8000000000000000;	/* ignore the sign */
+	return ((ieee754.u >> 52) == 0x7FF) &&
+		((ieee754.u & 0x000fffffffffffff) != 0);
+}
+#endif
+
+#ifndef	isinff
+static int isinff(float x)
+{
+	union { uint32_t u; float f; } ieee754;
+	ieee754.f = x;
+	ieee754.u &= ~0x80000000;		/* ignore the sign */
+	return ((ieee754.u >> 23) == 0xFF) &&
+		((ieee754.u & 0x7FFFFF) == 0);
+}
+#endif
+
+#ifndef	isnanf
+static int isnanf(float x)
+{
+	union { uint32_t u; float f; } ieee754;
+	ieee754.f = x;
+	ieee754.u &= ~0x80000000;		/* ignore the sign */
+	return ((ieee754.u >> 23) == 0xFF) &&
+		((ieee754.u & 0x7FFFFF) != 0);
+}
+#endif
+
+/* small errors are expected, computed as percentage error */
+#define MAX_FLOAT_ERROR_PERCENT		(3.5e-5)
+#define MAX_DOUBLE_ERROR_PERCENT	(4.5e-14)
+
+ZTEST(test_c_lib, test_sqrtf)
+{
+int i;
+float	exponent;
+float	resf;
+double	error;
+
+
+	/* test the special cases of 0.0, NAN, -NAN, INF, -INF,  and -10.0 */
+	zassert_true(sqrtf(0.0f) == 0.0f, "sqrtf(0.0)");
+	zassert_true(isnanf(sqrtf(NANF)), "sqrt(nan)");
+	zassert_true(isnanf(sqrtf(-NANF)), "sqrtf(-nan)");
+	zassert_true(isinff(sqrtf(INFF)), "sqrt(inf)");
+	zassert_true(isnanf(sqrtf(-INFF)), "sqrt(-inf)");
+	zassert_true(isnanf(sqrtf(-10.0f)), "sqrt(-10.0)");
+
+	for (exponent = 1.0e-10f; exponent < 1.0e10f; exponent *= 10.0f) {
+		for (i = 0; i < NUM_TEST_FLOATS; i++) {
+			resf = sqrtf(test_floats[i] * exponent);
+			zassert_true((resf > 0.0f) && (resf < INFF),
+					"sqrtf out of range");
+			if ((resf > 0.0f) && (resf < INFF)) {
+				error = (test_floats[i] * exponent -
+						resf * resf) /
+					(resf * resf) * 100;
```

Comment:
I'd use the input value rather than the output value for the divisor here, that way a really large erroneous return value won't be masked.

Reply:
Sorry for the delay getting back on this. I has waffled between the two divisor possibilities, you have given me a good reason to pick one over the other.